### PR TITLE
partial correlation and simple sns pair plot

### PR DESCRIPTION
**Pull request by jameschapman19**
added in partial correlation and t test as in Dinga R, Schmaal L, Penninx BW, Veltman DJ, Marquand AF. Controlling for effects of confounding variables on machine learning predictions. BioRxiv. 2020 Jan 1.

There will definitely be mistakes in my process here. tests has a problem with test_estimator_API() but is nothing to do with anything I have done so not sure what the normal thing to do is as a contributor.

### CODE REVIEW

**Comment by jameschapman19** [comment]
Hang on I need to update docs!

**Comment by jameschapman19** [comment]
actually it's ok I did docstrings as I went I thought it would entail more than that

File: confounds/metrics.py [hunk 1]
```diff
@@ -6,3 +6,61 @@
...
+
+import numpy as np
+
+from confounds import Residualize
+from scipy import stats
+
+
+def partial_correlation(X, y=None):
+    """
+    Calculates the partial correlation
```
Comment (raamana) [inline confounds/metrics.py:18]: elaborate on what exact variables are used for computing the metric .. e.g. "computes partial correlation between ... and ... ", and say something about when it can be used e.g. top 2 applications

**Comment by raamana** [comment]
Thanks very much James for this PR. See my comments.

Also, would be converting / committing the script you showed us into a short tutorial? You can make a demo folder to add such scripts.. You know about Jupyter Notebooks right?

**Comment by jameschapman19** [comment]
Responded to your comment, added the demo with cca, split out the functions in a different way. One of those where it could be refactored in so many ways but this way means we can calculate pairwise t-statistics and p-values for partial correlations and then the prediction/target partial correlation is then just a wrapper around them.

**Comment by raamana** [comment]
Thanks James. I will get to it as soon as possible while attending OHBM :)

File: confounds/metrics.py [hunk 2]
```diff
@@ -6,3 +6,102 @@
...
+
+import numpy as np
+
+from confounds import Residualize
+from scipy import stats
+
+
+def partial_correlation(X, y=None):
+    """
+    Calculates the pairwise partial correlations between all of the variables in X with respect to some confounding variables y
```
Comment (raamana) [inline confounds/metrics.py:18]: Hi James, it is best to rename the confound variable to C or something other than [small] y! You mention how its different from the typical y in the ML literature, it is easily missed by many. How about using C or conf_var or something like that?

File: confounds/metrics.py [hunk 3]
```diff
@@ -6,3 +6,102 @@
...
+    -------
+    corr_p : ndarray
+        Returns the pairwise partial correlations of each variable in X
+    t_statistic : ndarray
+        Returns the associated t-statistics for these pairwise partial correlations
+    statistical_significance : ndarray
+        Returns the associated p-values for these pairwise partial correlations
+    """
+    corr_p = partial_correlation(X, y=y)
+    n = X.shape[0]
```
Comment (raamana) [inline confounds/metrics.py:62]: before you use the dimensions, we need to ensure they are inputted correctly, so rookie mistakes are avoided. Do you know of `check_array()` utils from sklearn?

File: confounds/metrics.py [hunk 4]
```diff
@@ -6,3 +6,102 @@
...
+    g = y.shape[1]
+    # Replace perfect correlations to ensure large but not infinite t statistic
+    corr_p[corr_p == 1] = 1 - 1e-7
+    # partial correlation degrees of freedom
+    df = n - 2 - g
+    t_statistic = corr_p * np.sqrt(df / (1 - corr_p ** 2))
```
Comment (raamana) [inline confounds/metrics.py:68]: is there a reference/citation you base this formula on? Can we mention it in the docs for this function?

File: confounds/metrics.py [hunk 5]
```diff
@@ -6,3 +6,102 @@
...
+    resx.fit(X, y)
+    deconfound_X = resx.transform(X, y)
+    return np.corrcoef(deconfound_X, rowvar=False)
+
+
+def partial_correlation_t_test(X, y=None):
+    """
+    Calculates the t-statistic and p-value for pairwise partial correlations between all of the variables in X with
+    respect to some confounding variables y.
+
```
Comment (raamana) [inline confounds/metrics.py:44]: it would be useful to mention when and how this would be used! When we get a chance we can write a jupyter notebook demo later, but mentioning of few examples would be helpful.

File: confounds/metrics.py [hunk 6]
```diff
@@ -6,3 +6,102 @@
...
+def prediction_partial_correlation(predictions, targets, confounds):
+    """
+    Returns the partial correlation between predictions and targets after residualizing the effect of confounds.
+    Also calculates the t-statistic and p-value for the statistical significance of this partial correlation which
+    is a measure of the predictive power of the model controlling for the effect of confounds.
+
+    References
+    -----------
+    Dinga R, Schmaal L, Penninx BW, Veltman DJ, Marquand AF. Controlling for effects of confounding variables on machine learning predictions. BioRxiv. 2020 Jan 1.
```
Comment (raamana) [inline confounds/metrics.py:81]: not sure if you use an IDE or not, but it would be great if you can follow PEP guidelines for coding style etc e.g. width no longer than 80 chars etc, that is more or standard in the open source community?

File: confounds/metrics.py [hunk 7]
```diff
@@ -6,3 +6,102 @@
...
+    statistical_significance = stats.t.sf(np.abs(t_statistic), df=df)
+    return corr_p, t_statistic, statistical_significance
+
+
+def prediction_partial_correlation(predictions, targets, confounds):
+    """
+    Returns the partial correlation between predictions and targets after residualizing the effect of confounds.
+    Also calculates the t-statistic and p-value for the statistical significance of this partial correlation which
+    is a measure of the predictive power of the model controlling for the effect of confounds.
+
```
Comment (raamana) [inline confounds/metrics.py:78]: few comments for the prev function apply to this one as well.

File: confounds/metrics.py (hunk 2 above)
Comment (jameschapman19) [reply confounds/metrics.py:18]: This one I wasn't sure even when writing what to use (I think I started with something like C) because partial correlation need not be with respect to confounds it could be any old variable. Happy to go with one of your suggestions as in the context of this package it is likely to always be confounds but just confirming that's what we want

Comment (raamana) [reply confounds/metrics.py:18]: let's try use or make terminology that is likely to be used majority of our users, and also make a small note that they don't have to be confounds, as it is a generic computation.

that said, we are making a new function because it can not easily computed with scipy/numpy, yes?

File: confounds/metrics.py (hunk 3 above)
Comment (jameschapman19) [reply confounds/metrics.py:77]: Done this (and also learnt a trick for my own work so thanks!)

File: confounds/metrics.py (hunk 2 above)
Comment (jameschapman19) [reply confounds/metrics.py]: Have gone with C in the context of this package

File: confounds/metrics.py (hunk 4 above)
Comment (jameschapman19) [reply confounds/metrics.py:83]: Dinga et al. make this really clear but it's essentially public domain. Have added the Dinga reference to this code section

**Comment by raamana** [comment]
Hi James, I noticed you added two more commits. I won't review them until you tell me you're done with making changes and it is ready for review. Please mention me here to ensure I won't miss that request. 

ps: also I'd try change the commit message for each commit to reflect the actual changes. something short and simple

**Comment by jameschapman19** [comment]
Ok changes done @raamana

File: confounds/metrics.py [hunk 8]
```diff
@@ -6,3 +6,119 @@
...
+from scipy import stats
+from sklearn.utils.validation import check_array
+
+from confounds import Residualize
+
+
+def partial_correlation(X, C=None):
+    """
+    Calculates the pairwise partial correlations between all of the variables in X with respect to some confounding
+    variables y
```
Comment (raamana) [inline confounds/metrics.py:20]: typo y must be C

File: confounds/metrics.py (hunk 5 above)
Comment (raamana) [reply confounds/metrics.py:52]: @jameschapman19 few sentences for this one?

File: confounds/metrics.py (hunk 4 above)
Comment (raamana) [reply confounds/metrics.py:83]: Sure! but it does help to summarize your understanding in few sentences so users are not redirected! scikit-learn docs are great in that way as they tend more or less self-sufficient when possible. Do what you can.

File: confounds/visualize.py [hunk 9]
```diff
@@ -23,8 +24,7 @@ def samplets(features,
 
 
 def pairs(confounds,
```
Comment (raamana) [inline confounds/visualize.py:26]: Hi @jameschapman19 are you able to share few example visualizations to add to our docs website? esp for a differing number of variables (n>3)..

File: confounds/tests/test_confounds.py [hunk 10]
```diff
@@ -68,6 +68,21 @@ def test_residualize_linear():
...
+    train_X, train_confounds = splitter_X_confounds(train_all, 1)
+    # check that partial correlation with no confounds is the same as correlation using np.corrcoef
+    assert_almost_equal(np.corrcoef(train_X, rowvar=False),
+                        partial_correlation(train_X, np.zeros((train_X.shape[0], 1))))
+    # check that a linear regression fit using all variables has a lower r**2 partial correlation.
+    lr = LinearRegression().fit(train_all, train_y)
+    pred = lr.predict(train_all)
+    corr_p, t_stat, p_val = prediction_partial_correlation(pred,train_y,train_confounds)
+    assert_array_less(lr.score(train_all,train_y),corr_p)
+
```
Comment (raamana) [inline confounds/tests/test_confounds.py:84]: don't we need another test for partial_correlation_t_test?

**Review by raamana — CHANGES_REQUESTED** [review]

**Comment by raamana** [comment]
hi @jameschapman19, if you dont feel another test is necessary (as I noted above), let me know, so we can discuss and I can merge it asap. Cheers :)

**Comment by jameschapman19** [comment]
Just confirming I have seen this and will look in next couple of days

**Comment by raamana** [comment]
thanks James! take your time.

**Comment by raamana** [comment]
How goes James?

**Comment by jameschapman19** [comment]
Just looking at this now/today

**Comment by jameschapman19** [comment]
Think I have addressed your comments

**Comment by raamana** [comment]
Thanks James! happy holidays.

if you have other ideas or suggestions, please feel free to reach out anytime.
